### PR TITLE
Implement sanctums

### PR DESCRIPTION
**Pull request by htrajan**
Output of showCardInfo:

Sanctum of All {W}{U}{B}{R}{G}
At the beginning of your upkeep, you may search your library and/or graveyard for a Shrine card and put it onto the battlefield. If you search your library this way, shuffle it.
If an ability of another Shrine you control triggers while you control six or more Shrines, that ability triggers an additional time.

Sanctum of Calm Waters {3}{U}
At the beginning of your precombat main phase, you may draw X cards, where X is the number of Shrines you control. If you do, discard a card.

Sanctum of Stone Fangs {1}{B}
At the beginning of your precombat main phase, each opponent loses X life and you gain X life, where X is the number of Shrines you control.

Sanctum of Fruitful Harvest {2}{G}
At the beginning of your precombat main phase, add X mana of any one color, where X is the number of Shrines you control.

Sanctum of Shattered Heights {2}{R}
{1}, Discard a land card or Shrine card: Sanctum of Shattered Heights deals X damage to target creature or planeswalker, where X is the number of Shrines you control.

Sanctum of Tranquil Light {W}
{5}{W}: Tap target creature. This ability costs {1} less to activate for each Shrine you control.

### CODE REVIEW

File: Mage.Sets/src/mage/cards/s/SanctumOfAll.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+        super(ownerId, setInfo, new CardType[]{CardType.ENCHANTMENT}, "{W}{U}{B}{R}{G}");
+        
+        this.addSuperType(SuperType.LEGENDARY);
+        this.subtype.add(SubType.SHRINE);
+
+        // At the beginning of your upkeep, you may search your library and/or graveyard for a Shrine card and put it onto the battlefield. If you search your library this way, shuffle it.
+        this.addAbility(new BeginningOfUpkeepTriggeredAbility(new SanctumOfAllSearchEffect(), TargetController.YOU, true));
+
+        // If an ability of another Shrine you control triggers while you control six or more Shrines, that ability triggers an additional time.
+        this.addAbility(new SimpleStaticAbility(new SanctumOfAllTriggerEffect()).addHint(new ValueHint("Shrines you control", count)));
```
Comment (htrajan) [inline Mage.Sets/src/mage/cards/s/SanctumOfAll.java:57]: This should only copy triggered effects right? Does not copy the activated abilities of the white or red shrines.

File: Mage/src/main/java/mage/abilities/effects/common/cost/DynamicValueCostModificationEffect.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import mage.constants.CostModificationType;
+import mage.constants.Duration;
+import mage.constants.Outcome;
+import mage.game.Game;
+import mage.players.Player;
+import mage.util.CardUtil;
+
+import java.util.UUID;
+
+public class DynamicValueCostModificationEffect extends CostModificationEffectImpl {
```
Comment (htrajan) [inline Mage/src/main/java/mage/abilities/effects/common/cost/DynamicValueCostModificationEffect.java:15]: can apply this elsewhere later (such as in Pteramander)

**Comment by htrajan** [comment]
Looks like VerifyCardDataTest.verifyCards is failing. Anything I need to do? Failures seem unrelated.

File: Mage/src/main/java/mage/abilities/effects/common/cost/DynamicValueCostModificationEffect.java (hunk 2 above)
Comment (theelk801) [reply Mage/src/main/java/mage/abilities/effects/common/cost/DynamicValueCostModificationEffect.java:15]: does this effect still apply if something else copies the ability?

**Comment by htrajan** [comment]
Hmm I'm not sure. I followed the pattern that other such cost reduction effects use of checking the "original ID". Not sure what the intended behavior needs to be in this case but it should be consistent with what is out there (e.g., Pteramander).On Jun 24, 2020 7:08 AM, Evan Kranzler <notifications@github.com> wrote:
@theelk801 commented on this pull request.



In Mage/src/main/java/mage/abilities/effects/common/cost/DynamicValueCostModificationEffect.java:
> @@ -0,0 +1,64 @@
+package mage.abilities.effects.common.cost;
+
+import mage.abilities.Ability;
+import mage.abilities.Mode;
+import mage.abilities.dynamicvalue.DynamicValue;
+import mage.constants.CostModificationType;
+import mage.constants.Duration;
+import mage.constants.Outcome;
+import mage.game.Game;
+import mage.players.Player;
+import mage.util.CardUtil;
+
+import java.util.UUID;
+
+public class DynamicValueCostModificationEffect extends CostModificationEffectImpl {

does this effect still apply if something else copies the ability?

File: Mage/src/main/java/mage/abilities/effects/common/cost/DynamicValueCostModificationEffect.java (hunk 2 above)
Comment (htrajan) [reply Mage/src/main/java/mage/abilities/effects/common/cost/DynamicValueCostModificationEffect.java:15]: Hmm I'm not sure. I followed the pattern that other such cost reduction effects use of checking the "original ID". Not sure what the intended behavior needs to be in this case but it should be consistent with what is out there (e.g., Pteramander)

Comment (JayDi85) [reply Mage/src/main/java/mage/abilities/effects/common/cost/DynamicValueCostModificationEffect.java:15]: Must follow name rules: `CostModificationSourceEffect`:
* no need in `DynamicValue` prefix;
* `SourceEffect` -- that's effect for source object and without target;
* BTW: `xxxTargetEffect` -- with targets; `xxxControlledEffect` for own objects only; `xxxAllEffect` -- for any objects.

File: Mage.Sets/src/mage/cards/s/SanctumOfAll.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        super(card);
+    }
+
+    @Override
+    public SanctumOfAll copy() {
+        return new SanctumOfAll(this);
+    }
+}
+
+class SanctumOfAllSearchEffect extends OneShotEffect {
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/s/SanctumOfAll.java:70]: There is already exists effect: `SearchLibraryGraveyardPutOntoBattlefieldEffect`. I updated it last commit 6dfacfbca5eff906072cfee7a0739c52b0b498ed, so you can you use it here.

File: Mage.Sets/src/mage/cards/s/SanctumOfAll.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+            }
+            return true;
+        }
+        return false;
+    }
+}
+
+class SanctumOfAllTriggerEffect extends ReplacementEffectImpl {
+
+    private static final FilterPermanent filter = new FilterPermanent();
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/s/SanctumOfAll.java:129]: Miss controlled filter

File: Mage.Sets/src/mage/cards/s/SanctumOfStoneFangs.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        
+        this.addSuperType(SuperType.LEGENDARY);
+        this.subtype.add(SubType.SHRINE);
+
+        // At the beginning of your precombat main phase, each opponent loses X life and you gain X life, where X is the number of Shrines you control.
+        Ability ability = new BeginningOfPreCombatMainTriggeredAbility(
+                new LoseLifeOpponentsEffect(new PermanentsOnBattlefieldCount(filter, null)).setText("each opponent loses X life"),
+                TargetController.YOU, false)
+                .addHint(new ValueHint("Shrines you control", count));
+        ability.addEffect(new GainLifeEffect(new PermanentsOnBattlefieldCount(filter)).setText("and you gain X life, where X is the number of Shrines you control"));
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/s/SanctumOfStoneFangs.java:44]: Can be replaced by `count` in both places. Also I recommends names x values as `xValue` or another, but not `count` (count confuses).

File: Mage/src/main/java/mage/abilities/effects/common/cost/DynamicValueCostModificationEffect.java (hunk 2 above)
Comment (htrajan) [reply Mage/src/main/java/mage/abilities/effects/common/cost/DynamicValueCostModificationEffect.java]: fixed

File: Mage.Sets/src/mage/cards/s/SanctumOfAll.java (hunk 3 above)
Comment (htrajan) [reply Mage.Sets/src/mage/cards/s/SanctumOfAll.java]: fixed (also fixed a spacing issue in the staticText of `SearchLibraryGraveyardPutOntoBattlefieldEffect`)

File: Mage.Sets/src/mage/cards/s/SanctumOfAll.java (hunk 4 above)
Comment (htrajan) [reply Mage.Sets/src/mage/cards/s/SanctumOfAll.java]: actually didn't need that filter anymore so removed it

File: Mage.Sets/src/mage/cards/s/SanctumOfStoneFangs.java (hunk 5 above)
Comment (htrajan) [reply Mage.Sets/src/mage/cards/s/SanctumOfStoneFangs.java]: fixed

File: Mage/src/main/java/mage/abilities/effects/common/cost/DynamicValueCostModificationEffect.java (hunk 2 above)
Comment (JayDi85) [reply Mage/src/main/java/mage/abilities/effects/common/cost/DynamicValueCostModificationEffect.java]: I've got that @theelk801 talking about. `originalId` is smell code and can be broken for copied abilities/cards. See #5499 for more details and fixes example (compare source). Please, try that fix in PR.

Comment (htrajan) [reply Mage/src/main/java/mage/abilities/effects/common/cost/DynamicValueCostModificationEffect.java]: ah, gotcha, fixed following the example here: https://github.com/magefree/mage/commit/cae517e3105f9e3f2557c72dade71e6bedcc0f48

**Pull request merged**